### PR TITLE
Add a new 'close_tree' plugin to close entire trees by closing the pa…

### PR DESCRIPTION
**Pull request by redstreet**
Add a new 'close_tree' plugin to close entire trees by closing the parent.

See https://groups.google.com/g/beancount/c/c0Fm4hrxpUQ

### CODE REVIEW

File: beancount/plugins/close_tree.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    for entry in new_entries:
+        if isinstance(entry, Open) and (entry.account, entry.date) in closes:
+            continue
+        if isinstance(entry, Close) and (entry.account, entry.date) in opens:
+            continue
+        retval.append(entry)
+
+    if DEBUG:
+        elapsed_time = time.time() - start_time
+        print("Close account tree [{:.2f}s]: {} close entries added.".format(elapsed_time, close_count))
```
Comment (dnicolodi) [inline beancount/plugins/close_tree.py]: From this I infer that you are worried about the execution time of this plugin. If this is the case, maybe storing the accounts tree in a flat list is not the best choice. You could use a tree data structure. Also iterating the list of entries six times is not optimal. If there are no performance issues, I would delete the debug code, as it is anyhow not discoverable or accessible by the users and there are better profiling tools.

File: beancount/plugins/close_tree.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    # 2021-06-17 close Expenses:Non-Retirement:Auto:Fit
+    # plugin "beancount.plugins.close_tree"
+    # 2019-01-01 * "Transaction"
+    #   Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
+    #   Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
+    #   Assets:Transfer
+
+    # Above, auto_accounts inserts an Open for 'Expenses:Non-Retirement:Auto:Fit:Gas' on 2021-06-17
+    # since it appears in a close directive. But that's not what we want. The semantic of the close
+    # when u used with this plugin is: "close all children, and close the parent if it was opened
```
Comment (dnicolodi) [inline beancount/plugins/close_tree.py]: Spurious `u`.

File: beancount/plugins/close_tree.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                new_entries.append(entry)
+        else:
+            new_entries.append(entry)
+
+    # beancount.plugins.auto_accounts, if used in conjunction with this plugin, inserts an Open
+    # entry for empty parent accounts which we are trying to close. For example:
+
+    # plugin "beancount.plugins.auto_accounts"
+    # 2021-06-17 close Expenses:Non-Retirement:Auto:Fit
+    # plugin "beancount.plugins.close_tree"
```
Comment (dnicolodi) [inline beancount/plugins/close_tree.py]: I would put both `plugin` directives at the beginning.

File: beancount/plugins/close_tree.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    # 2019-01-01 * "Transaction"
+    #   Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
+    #   Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
+    #   Assets:Transfer
+
+    # Above, auto_accounts inserts an Open for 'Expenses:Non-Retirement:Auto:Fit:Gas' on 2021-06-17
+    # since it appears in a close directive. But that's not what we want. The semantic of the close
+    # when u used with this plugin is: "close all children, and close the parent if it was opened
+    # earlier. Below, we filter out entries to correctly handle this scenario:
```
Comment (dnicolodi) [inline beancount/plugins/close_tree.py]: I had a very hard time to follow what you are doing here. I had to unpack the different steps. Starting with this ledger:
```
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
the `bencount.plugins.auto_accounts` adds the `open` directives:
```
2019-01-01 open Assets:Transfer
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 open Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
Note that this this generates an error, because you have the open and the close directive on the same day, and this is not currently allowed. Let's ignore this for the moment. The `beancount.plugins.close_tree` adds the missing `close` directives based on the `open` directives present in the ledger:
```
2019-01-01 open Assets:Transfer
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 open Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Insurance
2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Gas
```
What you want to do is to remove the pair of `open` and `close` directives that appear on the same day. But waht if there is also a transaction to the relative account on the same day? You would need to keep the `open` and `close` directives. I think that the error message for `open` and `close` on the same day is not correct: Beancount puts `open` directives at the beginning of day and `close` directives at the end of day, and transactions in between. Thus I don't see a problem with `open` and `close` on the same day.

**Review by dnicolodi — CHANGES_REQUESTED** [review]

File: beancount/plugins/close_tree.py (hunk 3 above)
Comment (blais) [reply beancount/plugins/close_tree.py]: +1

File: beancount/plugins/close_tree.py (hunk 2 above)
Comment (blais) [reply beancount/plugins/close_tree.py]: +1

File: beancount/plugins/close_tree.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    close_count = 0
+    new_entries = []
+    errors = []
+
+    opens = [e.account for e in entries if isinstance(e, Open)]
+    closes = [e.account for e in entries if isinstance(e, Close)]
+
+    for entry in entries:
+        if isinstance(entry,  Close):
+            subaccounts = [a for a in opens if entry.account + ':' in a and a not in closes]
```
Comment (blais) [inline beancount/plugins/close_tree.py]: Use a set for speed

Comment (blais) [inline beancount/plugins/close_tree.py]: a.startswith(entry.account+ ":") or use a regexp

File: beancount/plugins/close_tree.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            for s in subaccounts:
```
Comment (blais) [inline beancount/plugins/close_tree.py]: Don't use single letter variables; rename to "account" or "sub" or something.

File: beancount/plugins/close_tree.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    Returns:
+      A tuple of entries and errors. """
+
+    start_time = time.time()
+    close_count = 0
+    new_entries = []
+    errors = []
+
+    opens = [e.account for e in entries if isinstance(e, Open)]
+    closes = [e.account for e in entries if isinstance(e, Close)]
```
Comment (blais) [inline beancount/plugins/close_tree.py]: If you wanted to be more precise about closing really just the subaccounts that are open at the time of closing, you should keep track of the tree of open accounts as you process the directives and close the subaccounts that are open then. You'd also have to handle subsequent closing directives colliding with your inserted ones. But I think that would be the most correct algorithm.

**Review by blais — CHANGES_REQUESTED** [review]

File: beancount/plugins/close_tree.py (hunk 7 above)
Comment (blais) [reply beancount/plugins/close_tree.py]: To be more clear what I mean is that it's possible for accounts to be opened, closed and then reopened (I think, I forget a bit).

File: beancount/plugins/close_tree.py (hunk 5 above)
Comment (redstreet) [reply beancount/plugins/close_tree.py:57]: Speed is not a concern, so leaving this as such (technically, I believe a list might be slightly faster in this case as the hashing of sets is avoided, and we don't do the expensive lookups or removals).

Comment (redstreet) [reply beancount/plugins/close_tree.py]: Changed.

File: beancount/plugins/close_tree.py (hunk 7 above)
Comment (redstreet) [reply beancount/plugins/close_tree.py]:
```
2014-01-01 open Assets:XBank
2015-01-01 close Assets:XBank

2016-01-01 open Assets:XBank
2017-01-01 close Assets:XBank   
```

This doesn't seem to be supported (in v2, at least). I get:
```
Duplicate open directive for Assets:XBank
Duplicate close directive for Assets:XBank
```

Do we still need to address this case? Is there a plan to support it in v3 or in the future?

File: beancount/plugins/close_tree.py (hunk 4 above)
Comment (redstreet) [reply beancount/plugins/close_tree.py]: 1. Fixed Beancount's error message.
2. Do we need to remove the pair, if there wasn't a transaction found to the account in question on that date? We would save the potentially expensive cost of checking if a removal is needed, if we instead always inserted the `open`/`close` pair (and sometimes ended up with this additional pair).

File: beancount/ops/validation.py [hunk 8]
```diff
@@ -87,7 +87,7 @@ def validate_open_close(entries, unused_options_map):
             else:
                 try:
                     open_entry = open_map[entry.account]
-                    if entry.date <= open_entry.date:
+                    if entry.date < open_entry.date:
```
Comment (dnicolodi) [inline beancount/ops/validation.py:90]: This should be a separate commit with a commit message that explain what is being changed and why. That said, I haven't checked if there is a reason for not allowing `open` and `close` directives on the same day. @blais do you remember anything about this?

File: beancount/plugins/close_tree.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+```
+
+Invoke this plugin _after_ any plugins that generate `open` directives for account trees that you
+want to auto close. An example is the `auto_accounts` plugin that ships with Beancount:
+
+```
+plugin "beancount.plugins.auto_accounts"
+plugin "beancount.plugins.close_tree"
+```
+"""
```
Comment (dnicolodi) [inline beancount/plugins/close_tree.py:24]: Documentation strings do not use Markdown syntax. If anything it should be reST.

File: beancount/plugins/close_tree.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def close_tree(entries, unused_options_map):
+    """Insert close entries for all subaccounts of a closed account.
+
+    Args:
+      entries: A list of directives. We're interested only in the Open/Close instances.
+      unused_options_map: A parser options dict.
+    Returns:
+      A tuple of entries and errors. """
```
Comment (dnicolodi) [inline beancount/plugins/close_tree.py]: The closing `"""` should go on a line by themselves.

File: beancount/plugins/close_tree_test.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+            2014-01-01 open Assets:XBank
+            2014-01-01 open Assets:XBank:AAPL
+            2015-01-01 close Assets:XBank
+            2015-01-01 close Assets:XBank:AAPL
+        """, dedent=True)
+
+        actual, _ = close_tree.close_tree(entries, {})
+        for a, e in zip(s(actual), s(expected)):
+            self.assertEqual(a.date, e.date)
+            self.assertEqual(a.account, e.account)
```
Comment (dnicolodi) [inline beancount/plugins/close_tree_test.py]: Here and everywhere else, is there a reason why `self.assertEqual(actual, expected)` does not work? Also, why is there the need to sort the actual and expected entries, can't the expected entries be in the right order?

**Review by dnicolodi — CHANGES_REQUESTED** [review]

File: beancount/plugins/close_tree_test.py (hunk 11 above)
Comment (redstreet) [reply beancount/plugins/close_tree_test.py]: 1. Doesn't work because metadata is not the same
2. Sorting the expected entries as the actual would have been sorted makes it unintuitive IMO.
```
            2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
            2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
            ....
            2021-06-17 open Expenses:Non-Retirement:Auto:Fit
            2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Insurance
            2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Gas
            2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
The above is neither alphabetical, nor consistent (ordering of opens and closes), nor does the open/close pair intuitively appear next to each other. In addition, changes to the plugin's algorithm might require all expected cases to be reordered, which is messy.

File: beancount/plugins/close_tree.py (hunk 7 above)
Comment (redstreet) [reply beancount/plugins/close_tree.py]: Ah, found the [documentation](https://beancount.github.io/docs/beancount_language_syntax.html#directives):
> _At the moment, once an account is closed, you cannot reopen it after that date._

@blais, do you still see this to be a point of concern?

File: beancount/plugins/close_tree_test.py (hunk 11 above)
Comment (dnicolodi) [reply beancount/plugins/close_tree_test.py]: Please have a look, for example, at how the tests for the `auto_accounts.py` plugin are structured and follow the same pattern.

File: beancount/ops/validation.py (hunk 8 above)
Comment (blais) [reply beancount/ops/validation.py:90]: I don't remember.
Il'l review now.

Comment (blais) [reply beancount/ops/validation.py:90]: It does make sense to me that we should allow entries on the same day as the open though. Accepting.

**Review by blais — APPROVED** [review]
